include/clang/AST/Expr.h

```diff
@@ -2798,7 +2798,9 @@ class BoundsExpr : public Expr {
 /// classes).
 class CastExpr : public Expr {
 private:
-  enum { OP, BOUNDS, END_EXPR = 2 };
+  // BOUNDS - enum value for bounds of cast expression
+  // SUBBOUNDS - enum value for bounds of subexpression
+  enum { OP, BOUNDS, SUBBOUNDS, END_EXPR = 3 };
```

Comment:
I noticed that the code is overwriting the bounds expression at BOUNDS during bounds inference.  This is to put it into the form of a range.   I don't think we should be overwriting the bounds expression at BOUNDS.   This means that the AST dump would not reflect what the user wrote, for example.   

I think there should be a new entry for the converted bounds too.